cmd/usecasevalidator/validators.go

```diff
@@ -0,0 +1,281 @@
+//go:build !nogrpcserver && !_test
+
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/ipfs/go-cid"
+	"github.com/samber/lo"
+
+	"github.com/anyproto/anytype-heart/core/block/editor/widget"
+	"github.com/anyproto/anytype-heart/core/block/simple"
+	"github.com/anyproto/anytype-heart/core/block/simple/bookmark"
+	"github.com/anyproto/anytype-heart/core/block/simple/dataview"
+	"github.com/anyproto/anytype-heart/core/block/simple/link"
+	"github.com/anyproto/anytype-heart/core/block/simple/relation"
+	"github.com/anyproto/anytype-heart/core/block/simple/text"
+	"github.com/anyproto/anytype-heart/core/domain"
+	"github.com/anyproto/anytype-heart/pb"
+	"github.com/anyproto/anytype-heart/pkg/lib/bundle"
+	"github.com/anyproto/anytype-heart/pkg/lib/localstore/addr"
+	"github.com/anyproto/anytype-heart/pkg/lib/pb/model"
+	"github.com/anyproto/anytype-heart/util/pbtypes"
+)
+
+type validator func(snapshot *pb.SnapshotWithType, info *useCaseInfo) error
+
+var validators = []validator{
+	validateRelationLinks,
+	validateRelationBlocks,
+	validateDetails,
+	validateObjectTypes,
+	validateBlockLinks,
+	validateFileKeys,
+	validateDeleted,
+	validateRelationOption,
+}
+
+func validateRelationLinks(s *pb.SnapshotWithType, info *useCaseInfo) (err error) {
+	id := pbtypes.GetString(s.Snapshot.Data.Details, bundle.RelationKeyId.String())
+	for _, rel := range s.Snapshot.Data.RelationLinks {
+		if bundle.HasRelation(rel.Key) {
+			continue
+		}
+		v, found := info.customTypesAndRelations[rel.Key]
+		if found {
+			v.isUsed = true
+			info.customTypesAndRelations[rel.Key] = v
+		} else {
+			err = multierror.Append(err, fmt.Errorf("object '%s' contains link to unknown relation: %s(%s)", id,
+				rel.Key, pbtypes.GetString(s.Snapshot.Data.Details, bundle.RelationKeyName.String())))
+		}
+	}
+	return err
+}
+
+func validateRelationBlocks(s *pb.SnapshotWithType, _ *useCaseInfo) (err error) {
+	id := pbtypes.GetString(s.Snapshot.Data.Details, bundle.RelationKeyId.String())
+	var relKeys []string
+	for _, b := range s.Snapshot.Data.Blocks {
+		if rb, ok := simple.New(b).(relation.Block); ok {
+			relKeys = append(relKeys, rb.Model().GetRelation().Key)
+		}
+	}
+	relLinks := pbtypes.RelationLinks(s.Snapshot.Data.GetRelationLinks())
+	for _, rk := range relKeys {
+		if !relLinks.Has(rk) {
+			err = multierror.Append(err, fmt.Errorf("relation '%v' exists in relation block but not in relation links of object %s", rk, id))
+		}
+	}
+	return err
+}
+
+func validateDetails(s *pb.SnapshotWithType, info *useCaseInfo) (err error) {
```

Comment:
I think, you should break it into several small functions

Reply:
Moved all logic related to collecting of custom objects' usage info to separate file